### PR TITLE
Introducing istioctl

### PR DESCRIPTION
**Pull request by kyessenov**
Many changes meant to improve usability of istioctl:
- add PUT/POST to config registry
- make istioctl standalone
- move ingress kind blacklisting down to client from controller
- add -f for file input
- always log to stderr
- update test to modify route rules
- add config flow doc

### CODE REVIEW

File: doc/configuration-flow.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# Overview of the configuration flow in Istio Manager
+
+Istio configuration consists of configuration objects. Each object has kind
+(for example, a routing rule), name and namespace uniquely identifying the
+object.  Istio Manager provides a REST-based interface for manipulating objects
```
Comment (mandarjog) [inline doc/configuration-flow.md]: This makes it sound like I can  do
`curl https://istio_manager/...  -X PUT -d "{}"`, if that is not the case this comment is misleading. 
If it *is* the case, we should show `curl` examples.

Comment (kyessenov) [inline doc/configuration-flow.md]: It is sort of the case, since `istioctl` essentially issues curl commands to kube API server. `curl` is difficult to show since we need to translate URL, add auth headers, and not do validation.

Comment (kyessenov) [inline doc/configuration-flow.md]: REST != HTTP, but I agree REST doesn't have a clear definition anymore.

**Comment by rshriram** [comment]
@kyessenov do we need to expose put/post semantics to end user? How about we just do istioctl create, istioctl update, istioctl get, istioctl delete ?

File: cmd/root.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	Client *kube.Client
+
+	// RootCmd is the root CLI command
+	RootCmd = &cobra.Command{
+		Short: "Istio Manager",
+		PersistentPreRunE: func(cmd *cobra.Command, args []string) (err error) {
+			glog.V(2).Infof("Root flags: %#v", RootFlags)
+			Client, err = kube.NewClient(RootFlags.Kubeconfig, model.IstioConfig)
+			if err != nil {
+				Client, err = kube.NewClient(os.Getenv("HOME")+"/.kube/config", model.IstioConfig)
```
Comment (rshriram) [inline cmd/root.go]: shouldn't this move to istioctl instead of staying here?

File: cmd/root.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+				return multierror.Prefix(err, "Failed to register Third-Party Resources.")
+			}
+			return
+		},
+	}
+)
+
+func init() {
+	RootCmd.PersistentFlags().StringVarP(&RootFlags.Kubeconfig, "kubeconfig", "c", "",
+		"Use a Kubernetes configuration file instead of in-cluster configuration")
```
Comment (rshriram) [inline cmd/root.go]: This as well.

File: doc/configuration-flow.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+platform-specific persistent store, and then triggers a notification event at
+the remote proxy component.
+
+## User input
+
+Istio Manager provides a command line interface called `istioctl` that exposes
+the REST-based configuration store interface.  The tool validates that its YAML
+input satisfies the schema for the registered Istio configuration kind, and
+stores the validated input in the store.
+
```
Comment (rshriram) [inline doc/configuration-flow.md]: This is good. It might need some massaging, in terms of kind->type, etc. But the watering down of the message can be done later or in another doc that is for public consumption while we keep this for internal use (design doc).

File: doc/configuration-flow.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+Istio configuration storage is platform-specific. Below, we focus on Kubernetes,
+since Istio Manager model is close to Kubernetes model of Third-Party
+Resources. In fact, `kubectl` could be used instead of `istioctl` if validation
+and patching semantics of Third-Party Resources was supported.
+
+Each Istio configuration object is stored as a Kubernetes Third-Party resource
+of kind `istioconfig`. The name of the resource is a combination of Istio kind
+and configuration name, and namespaces are shared between Kubernetes and Istio.
+Internally, this means Kubernetes assigns a key subset in `etcd` to Istio,
+exposes an HTTP endpoint for streaming updated to the store, and provides the
```
Comment (rshriram) [inline doc/configuration-flow.md]: nit: typo - s/updated/updates/

File: doc/configuration-flow.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+also allows watching for changes to the configuration store. For example, a
+function can subscribe to receive object creation events for objects of a
+particular kind. The controller streams the events to the subscribers, and also
+maintains a local cache view of the entire store.
+
+The flow of configuration starts at the operator, continues to the
+platform-specific persistent store, and then triggers a notification event at
+the remote proxy component.
+
+## User input
```
Comment (rshriram) [inline doc/configuration-flow.md]: Can you number the sections so that its easier to map to the flow of configuration?

File: platform/kube/client.go [hunk 7]
```diff
@@ -261,21 +261,49 @@ func (cl *Client) Get(key model.Key) (proto.Message, bool) {
 	return out, true
 }
 
-// Put implements registry operation
-func (cl *Client) Put(k model.Key, v proto.Message) error {
-	out, err := modelToKube(cl.mapping, &k, v)
+// Post implements registry operation
+func (cl *Client) Post(key model.Key, v proto.Message) error {
+	if key.Kind == model.IngressRule {
+		return fmt.Errorf("unsupported operation: cannot put a config element of kind '%s'", key.Kind)
```
Comment (rshriram) [inline platform/kube/client.go]: nit: cannot create Ingress rule through third party config
isn't that the intent?

File: test/integration/rule-content-route.yaml.tmpl [hunk 8]
```diff
@@ -1,5 +1,5 @@
 destination: "{{.destination}}.{{.namespace}}.svc.cluster.local"
-precedence: 3
+precedence: 2
```
Comment (rshriram) [inline test/integration/rule-content-route.yaml.tmpl]: why this change?

File: test/integration/driver.go [hunk 9]
```diff
@@ -232,28 +231,34 @@ func getRestartEpoch(pod string) (int, error) {
...
 		Name:      name,
 		Namespace: params.namespace,
-	}, v))
+	}
+	if create {
+		check(istioClient.Post(key, v))
+	} else {
+		check(istioClient.Put(key, v))
+	}
 }
```
Comment (frankbu) [inline test/integration/driver.go]: I wonder if instead of having to CLI commands (POST and PUT), there could just be one (ADD) with "add or create" semantics? Or all of the above, but maybe that's overkill.

File: doc/configuration-flow.md (hunk 1 above)
Comment (frankbu) [inline doc/configuration-flow.md]: A --debug option that displays the equivalent HTTP/curl request that the command is doing might be helpful for understanding

File: platform/kube/kube_test.go [hunk 10]
```diff
@@ -133,9 +133,9 @@ func TestControllerCacheFreshness(t *testing.T) {
 	k := model.Key{Kind: mock.Kind, Name: "test", Namespace: ns}
 	o := mock.Make(0)
 
-	// put followed by delete
-	glog.Infof("Calling Put(%#v)", k)
-	if err := ctl.Put(k, o); err != nil {
+	// add and remove
+	glog.Infof("Calling Postt(%#v)", k)
```
Comment (frankbu) [inline platform/kube/kube_test.go]: s/Postt/Post/

File: cmd/istioctl/main.go [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+
+	key    model.Key
+	schema model.ProtoSchema
+
+	postCmd = &cobra.Command{
+		Use:   "post [kind] [name]",
+		Short: "Create a configuration object",
+		RunE: func(c *cobra.Command, args []string) error {
+			if len(args) != 2 {
+				return fmt.Errorf("Provide kind and name")
```
Comment (ayj) [inline cmd/istioctl/main.go]: nit: lowercase error strings here and elsewhere unless the cobra command error handling doesn't prefix these errors with anything else (see https://github.com/golang/go/wiki/CodeReviewComments#error-strings)

**Comment by frankbu** [comment]
Not exactly an issue with this particular PR, but while looking at the doc part, I noticed that `proxy-injection.md` is not up to date. It's missing the new requirement to add POD_NAME/NAMESPACE/IP env variables.

File: cmd/root.go [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+	// Client is a kubernetes client interface set-up by the root command
+	Client *kube.Client
+
+	// RootCmd is the root CLI command
+	RootCmd = &cobra.Command{
+		Short: "Istio Manager",
+		PersistentPreRunE: func(cmd *cobra.Command, args []string) (err error) {
+			glog.V(2).Infof("Root flags: %#v", RootFlags)
+			Client, err = kube.NewClient(RootFlags.Kubeconfig, model.IstioConfig)
```
Comment (ayj) [inline cmd/root.go]: Consider not using named result parameters to avoid declaring err var inside the function (see https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters)

File: cmd/root.go [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+		"Use a Kubernetes configuration file instead of in-cluster configuration")
+	RootCmd.PersistentFlags().StringVarP(&RootFlags.Namespace, "namespace", "n", "",
+		"Select a Kubernetes namespace")
+
+	// carry over glog flags with new defaults
+	flag.CommandLine.VisitAll(func(gf *flag.Flag) {
+		switch gf.Name {
+		case "logtostderr":
+			err := gf.Value.Set("true")
+			// cannot handle error here since logging is not initialized
```
Comment (ayj) [inline cmd/root.go]: Can you use fmt.Printf here to print pre-log initialization errors?

File: platform/kube/kube_test.go [hunk 14]
```diff
@@ -268,9 +268,9 @@ func TestProxyConfig(t *testing.T) {
 
 	key := model.Key{Kind: model.RouteRule, Name: "test", Namespace: ns}
 
-	err := cl.Put(key, rule)
+	err := cl.Post(key, rule)
```
Comment (ayj) [inline platform/kube/kube_test.go]: (nit) These two lines could be combined, e.g.

if err := cl.Post(key, rule); err != nil {
    t.Errorf(...)
}

File: test/integration/driver.go (hunk 9 above)
Comment (rshriram) [inline test/integration/driver.go]: I think with just one create command, it would be slightly confusing to the end user. Especially, given that kubernetes has `kubectl apply` (which provides update semantics). In the interest of clarity, it might be useful to explicitly have create, update, get and delete. What do you think?

Comment (kyessenov) [inline test/integration/driver.go]: Yes, we need PATCH semantics to make "apply" work, and there the difficulty is dealing with lists in protos. That's the path kubectl went on, when a similar question was brought up (why have "create" and "replace"). It's also nice to keep each command to one atomic operation.

File: cmd/istioctl/main.go [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+
+var (
+	// input file name
+	file string
+
+	key    model.Key
+	schema model.ProtoSchema
+
+	postCmd = &cobra.Command{
+		Use:   "post [kind] [name]",
```
Comment (kyessenov) [inline cmd/istioctl/main.go]: done

File: cmd/istioctl/main.go [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+			v, err := readInput()
+			if err != nil {
+				return err
+			}
+			return cmd.Client.Post(key, v)
+		},
+	}
+
+	putCmd = &cobra.Command{
+		Use:   "put [kind] [name]",
```
Comment (kyessenov) [inline cmd/istioctl/main.go]: done

File: cmd/root.go (hunk 2 above)
Comment (kyessenov) [inline cmd/root.go]: it's shared with cmd/manager

File: cmd/root.go (hunk 3 above)
Comment (kyessenov) [inline cmd/root.go]: it's shared with cmd/manager/main.go

File: doc/configuration-flow.md (hunk 6 above)
Comment (kyessenov) [inline doc/configuration-flow.md]: done

File: doc/configuration-flow.md (hunk 5 above)
Comment (kyessenov) [inline doc/configuration-flow.md]: done

File: platform/kube/client.go (hunk 7 above)
Comment (kyessenov) [inline platform/kube/client.go]: yes, that's why we catch mutations here.

File: test/integration/rule-content-route.yaml.tmpl (hunk 8 above)
Comment (kyessenov) [inline test/integration/rule-content-route.yaml.tmpl]: testing PUT instead of POST. content and default route are named the same in the test driver, but the third rule uses precedence.

File: platform/kube/kube_test.go (hunk 10 above)
Comment (kyessenov) [inline platform/kube/kube_test.go]: done

File: cmd/istioctl/main.go (hunk 11 above)
Comment (kyessenov) [inline cmd/istioctl/main.go]: done, thanks

File: cmd/root.go (hunk 12 above)
Comment (kyessenov) [inline cmd/root.go]: I think it's ok here, since Client is a global variable.

File: cmd/root.go (hunk 13 above)
Comment (kyessenov) [inline cmd/root.go]: done

File: platform/kube/kube_test.go (hunk 14 above)
Comment (kyessenov) [inline platform/kube/kube_test.go]: done

File: doc/configuration-flow.md (hunk 1 above)
Comment (kyessenov) [inline doc/configuration-flow.md]: -v 10 will print kube API server curl commands, so this is already supported.

**Comment by kyessenov** [comment]
Thanks for the comments! I addressed them in the last commit.

File: platform/kube/client.go (hunk 7 above)
Comment (rshriram) [inline platform/kube/client.go]: No I meant, change the error message to "cannot create Ingress rule through third party resource config"

Comment (kyessenov) [inline platform/kube/client.go]: done
